### ISSUE TITLE
Declarative syntax in Python

### ISSUE THREAD

**Issue by jamesdbrock** [pyedifice/pyedifice#161]
The ideal syntax for declaring render trees in Edifice would be a declarative expression syntax.

But Python is stubbornly statement-oriented. It's almost impossible to branch inside expressions. If if-expression in Python is terrible, and [match-expression doesn't exist](https://peps.python.org/pep-0622/#make-it-an-expression).

So we currently declare render trees by using statements. This requires statements to use an implicit invisible render context.

```python
with View():
  Label("one")
    with View():
      Label("two")
      Label("three")
```

**Comment by jamesdbrock** [comment]
Maybe we could remove the implicit `Tracker` and instead just go back to expression syntax with the `__call__` method for render trees like this.

```python
return View()(
  Label("one"),
  *([View()(Label("two"), Label("three"))] if condition else [])
)
```

This syntax is trash, so we could introduce an *explicit* `RenderTreeContext` to help with declaring the trees.

```python
class RenderTreeContext():
  def __init__():
    self.stack = []
  def __call__(element:Element):
    self.stack[-1](element)
    return element
  def __enter__(element:Element):
    self.stack.append(element)
    return element
  def __exit__():
    return self.stack.pop()
```

Also we make `Element.__call__()` append to its children instead of setting its children.

Then users could explicitly use the `RenderTreeContext` to help declaring render trees.

```python
rtc = RenderTreeContext()

return with rtc(View()):
  rtc(Label("one"))
  if condition:
    with rtc(View()):
      rtc(Label("one"))
      rtc(Label("two"))
```

**Comment by considerate** [comment]
I would also like the `@component` to return the `Element` but I can't think of a good way to do so if we have render contexts with the `with` statement as you would end up with something like

```Python
with render(View()) as main_view:
    with render(View()):
         ...
    ...
    ...
    return main_view
```

Which I feel is bot non-intuitive and the `return` statement doesn't _feel_ like you're returning the full element tree of the component.

**Comment by jamesdbrock** [comment]
How about this.

```python
class Element():
    def __init__(self, label:str):
        self.label = label
        self.children = []
    def __call__(self, *args):
        self.children.extend(args)
        return self
    def __str__(self) -> str:
        return self.label + " [" + " ".join(map(str, self.children)) + "]"

class RenderTreeContext():
    def __init__(self):
        self.stack = []
    def __call__(self, element:Element):
        if len(self.stack) == 0:
            self.stack.append(element)
        else:
            self.stack[-1](element)
        return RenderTreeContextManager(self, element)
    def root(self):
        return self.stack[0]

class RenderTreeContextManager():
    def __init__(self, rtc: RenderTreeContext, element: Element):
        self.rtc = rtc
        self.element = element
    def __enter__(self):
        self.rtc.stack.append(self.element)
        return self.element
    def __exit__(self, exc_type, exc_value, traceback):
        self.rtc.stack.pop()
        return True


def component_render(condition: bool) -> Element:
    rtc = RenderTreeContext()
    with rtc(Element("main")):
        rtc(Element("one"))
        if condition:
            with rtc(Element("sub")):
                rtc(Element("two"))
                rtc(Element("three"))
    return rtc.root()

if __name__ == "__main__":

    print(Element("main")(Element("one"), Element("two")))

    print(component_render(True))
    print(component_render(False))
```

Output
```
main [one [] two []]
main [one [] sub [two [] three []]]
main [one []]
```

**Comment by jamesdbrock** [comment]
Gonna call this thing `TreeBuilder` and use it like:

```python
put = TreeBuilder()
```

**Comment by jamesdbrock** [comment]
Also going to return from inside the `with` like:

```python
put = TreeBuilder()
  with put(View()) as root:
    put(Label())
    put(Label())
    ...
    return root

**Comment by considerate** [comment]
Half-baked idea:

Can we somehow combine the `__call__` style with the `with` style?

I'm thinking something along these lines:

```Python

def complicated_tree():
     with put(View()) as root:
         for text in texts:
             put(Label(text))
         return root

return View(Label("first"), Delayed(complicated_tree), Label("third"))
```

Then the `__call__` method of an element would do something like:
```Python
def __call__(self, *args) -> Self:
     with put(self):
          for elem in args:
               match elem:
                    case Delayed(f):
                          f()
                    case _:
                          self.children.append(elem)
```

**Comment by considerate** [comment]
We don't actually need any TreeBuilder...

```Python
def complicated_tree():
    with View() as root:
        for text in texts:
            root(Label(text))
        return root

with View() as root:
  if has_error:
     root(Label("asd"))

  return root(
    Label("asd"),
    View()(
      Label("first"),
      complicated_tree(),
      Label("third")
    )
  )
```

**Comment by JoanCoCo** [comment]
I personally feel that all of these syntax changes make things somewhat more complex/confusing. One of the big strengths of Edifice is how easy is to build UIs and learn to use it, and I feel here we are walking the thin line about to loose this easiness.

Also, I do not understand what is the issue in keeping the invisible `Trackers`. Could please explain?

**Comment by jamesdbrock** [comment]
> I personally feel that all of these syntax changes make things somewhat more complex/confusing. One of the big strengths of Edifice is how easy is to build UIs and learn to use it, and I feel here we are walking the thin line about to loose this easiness.

Yeah.

> Also, I do not understand what is the issue in keeping the invisible `Trackers`. Could please explain?

- Current system: Element constructor with `Tracker` side-effect.
  Element constructors implicitly add themselves to the element tree as a side-effect when they are constructed.
  Nice-looking syntax, but then Elements don't behave like normal Python objects and the Edifice implementation for `Tracker` is complicated.

- System merged in #154 : `Element.render()` method which adds to `Tracker`.
  Now Elements behave like normal Python objects. There is still a `Tracker` and the Element gets added to the element tree by `Element.render()`. This is nice but the syntax is always kind of noisy and there's always a loose coupling between the Python syntax tree and the declared element tree, because you can call `.render()` whenever you want.
  
- System in development #162 : No `Tracker`.
  We go back to the original Edifice declaration syntax where you add children to an Element with `Element.__call__()`.
  So now Elements are normal objects and the Python expression syntax tree declares the element tree. Which is great but because Python expression syntax is so weak, it's hard to conditionally declare trees. This was what motivated us to first introduce the implicit `Tracker`. So for the case where you want a complicated conditional element tree, we introduce `TreeBuilder`, which is a lot like the `Tracker`, but explicit. `TreeBuilder` has noisy syntax, but you don't have to use it all the time.

**Comment by JoanCoCo** [comment]
> but you don't have to use it all the time.

I do not agree in that one. I think almost any more or less normal project will need conditionally rendering some elements or others.

**Comment by JoanCoCo** [comment]
I am not a fan of having to make the user decide when/if they should be using the `TreeBuilder`.

**Comment by jamesdbrock** [comment]
> I am not a fan of having to make the user decide when/if they should be using the `TreeBuilder`.

Yeah. As @considerate has pointed out to me, the user can get in a situation where they are writing a big Python expression to declare the element tree, and then down at the leaf they realize that they need a complicated condition, so then they have to go back up and change the whole Python expression tree to a `TreeBuilder` statement tree.

**Comment by jamesdbrock** [comment]
Notes for docs:

Python statements can contain expressions, but Python expressions cannot contain statements.

Python branching statements are first-class

- `if`
- `for`
- `match`

Python branching expressions are awkward and limited

- `if` Weird infix syntax.
- `for` Unpack with `*` operator.
- `match` Doesn't exist.

**Comment by jamesdbrock** [comment]
There is a Python docs article about functional programming https://docs.python.org/3/howto/functional.html

**Comment by jamesdbrock** [comment]
There is a React docs article about Conditional Rendering https://react.dev/learn/conditional-rendering

**Comment by considerate** [comment]
@jamesdbrock JavaScript has conditional expressions though.

**Comment by jamesdbrock** [comment]
> @jamesdbrock JavaScript has conditional expressions though.

You mean the ternary condition operator?

**Comment by jamesdbrock** [comment]
Reacton does almost exactly what we are currently doing with system **1.**

https://github.com/widgetti/reacton/blob/16518bf6fe9ca018bf8576398deb9c8bc27d6dc3/reacton/core.py#L337-L342

**Comment by considerate** [comment]
Yeah, our implementation is based on [nicegui](https://nicegui.io) and reacton.

**Comment by jamesdbrock** [comment]
One interesting thing about Reacton is that it requires the `@component` functions to return an `Element`.

https://github.com/widgetti/reacton/blob/master/notebooks/todo-app.ipynb

So the top-level `Element` is usually named by `with`..`as` and returned, and the inner `Element`s are implicit children.

```python
    with rv.ListItem() as main:
        with rv.Btn(icon=True):
            rv.Icon(children=["mdi-delete"])
        rv.TextField(v_model=item.text, on_v_model=on_change_text)
    return main
```

This is kind of a nice point in the design space.

But it doesn't help us with our  `children` props problem.

**Comment by hellozyemlya** [comment]
Is there considerations to create py('x' is already used, maybe 'm') file extension like jsx\tsx counterpart?

**Comment by jamesdbrock** [comment]
> Is there considerations to create py('x' is already used, maybe 'm') file extension like jsx\tsx counterpart?

No. We prefer to use Python as the declarative syntax. Inventing a new language for declaring widget trees would just result in a declarative language that is only marginally better than Python, but without all of the Python language ecosystem support.

That being said, there many other projects that are trying to invent a new declarative language for Python + Qt. Qt's QML is the obvious one. Another one is https://github.com/fork-tongue/collagraph
